File: filechooser.cpp
#include "filechooser.h"
#include <systemdispatcher.h>

const QString FileChooser::ContentChoosenMessage = QStringLiteral("AndroidUtils.FileChooser.contentChoosen");

FileChooser::FileChooser(QObject *parent) :
	QObject(parent),
	_title(),
	_contentUrl(),
	_type(OpenDocument),
	_mimeType(QStringLiteral("*/*")),
	_flags(OpenableFlag)
{
	auto dispatcher = AndroidNative::SystemDispatcher::instance();
	dispatcher->loadClass(QStringLiteral("de.skycoder42.androidutils.FileChooser"));
	connect(dispatcher, &AndroidNative::SystemDispatcher::dispatched,
			this, &FileChooser::onDispatched,
			Qt::QueuedConnection);
}

QString FileChooser::title() const
{
	return _title;
}

QUrl FileChooser::contentUrl() const
{
	return _contentUrl;
}

FileChooser::ChooserType FileChooser::type() const
{
	return _type;
}

QString FileChooser::mimeType() const
{
	return _mimeType;
}

FileChooser::ChooserFlags FileChooser::chooserFlags() const
{
	return _flags;
}

void FileChooser::setContentUrl(QUrl contentUrl)
{
	if (_contentUrl == contentUrl)
		return;

	_contentUrl = contentUrl;
	emit contentUrlChanged(contentUrl);
}

void FileChooser::setType(FileChooser::ChooserType type)
{
	if (_type == type)
		return;

	_type = type;
	emit typeChanged(type);
}

void FileChooser::setMimeType(QString mimeType)
{
	if (_mimeType == mimeType)
		return;

	_mimeType = mimeType;
	emit mimeTypeChanged(mimeType);
}

void FileChooser::setChooserFlags(ChooserFlags chooserFlags)
{
	if (_flags == chooserFlags)
		return;

	_flags = chooserFlags;
	emit chooserFlagsChanged(chooserFlags);
}

void FileChooser::onDispatched(const QString &message, const QVariantMap &data)
{
	if(message == ContentChoosenMessage) {
		if(data.value(QStringLiteral("success"), false).toBool()) {
			_contentUrl = data.value(QStringLiteral("uri")).toString();
			emit contentUrlChanged(_contentUrl);
			emit accepted();
		} else
			emit rejected();
	}
}

void FileChooser::open()
{
	QString message;
	QVariantMap data;

	switch (_type) {
	case FileChooser::GetContent:
		message = QStringLiteral("AndroidUtils.FileChooser.getContent");
		data = {
			{"title", _title.isEmpty() ? tr("Open File") : _title},
			{"mime", _mimeType},
			{"openable", _flags.testFlag(OpenableFlag)},
			{"localOnly", _flags.testFlag(LocalOnlyFlag)},
			{"grantWrite", _flags.testFlag(AlwaysGrantWriteFlag)}
		};
		break;
	case FileChooser::OpenDocument:
		message = QStringLiteral("AndroidUtils.FileChooser.openDocument");
		data = {
			{"title", _title.isEmpty() ? tr("Open File") : _title},
			{"mime", _mimeType},
			{"url", _contentUrl.toString()},
			{"openable", _flags.testFlag(OpenableFlag)},
			{"grantWrite", _flags.testFlag(AlwaysGrantWriteFlag)},
			{"persistPermissions", _flags.testFlag(PersistPermissionsFlag)}
		};
		break;
	case FileChooser::CreateDocument:
		message = QStringLiteral("AndroidUtils.FileChooser.createDocument");
		data = {
			{"title", _title.isEmpty() ? tr("Save File") : _title},
			{"mime", _mimeType},
			{"url", _contentUrl.toString()},
			{"name", _contentUrl.fileName()},
			{"openable", _flags.testFlag(OpenableFlag)},
			{"persistPermissions", _flags.testFlag(PersistPermissionsFlag)}
		};
		break;
	default:
		Q_UNREACHABLE();
		break;
	}

	AndroidNative::SystemDispatcher::instance()->dispatch(message, data);
}

void FileChooser::setTitle(QString title)
{
	if (_title == title)
		return;

	_title = title;
	emit titleChanged(title);
}

File: android/src/de/skycoder42/androidutils/FileChooser.java
package de.skycoder42.androidutils;

import java.util.Map;
import java.util.HashMap;
import android.net.Uri;
import android.content.Context;
import android.content.Intent;
import android.content.ContentResolver;
import android.app.Activity;
import org.qtproject.qt5.android.QtNative;
import androidnative.SystemDispatcher;

public class FileChooser {
	private static final int CHOOSE_CONTENT_ACTION = 0x1091c657;
	private static final int CHOOSE_PERSISTENT_ACTION = 0x1091c658;

	private static final String GET_CONTENT_MESSAGE = "AndroidUtils.FileChooser.getContent";
	private static final String OPEN_DOCUMENT_MESSAGE = "AndroidUtils.FileChooser.openDocument";
	private static final String CREATE_DOCUMENT_MESSAGE = "AndroidUtils.FileChooser.createDocument";
	private static final String CONTENT_CHOOSEN_MESSAGE = "AndroidUtils.FileChooser.contentChoosen";

	static {
		SystemDispatcher.addListener(new SystemDispatcher.Listener() {
			public void onDispatched(String type , Map message) {
				if (type.equals(GET_CONTENT_MESSAGE)) {
					getContent((String)message.get("title"),
							   (String)message.get("mime"),
							   (Boolean)message.get("openable"),
							   (Boolean)message.get("localOnly"),
							   (Boolean)message.get("grantWrite"));
				} else if (type.equals(OPEN_DOCUMENT_MESSAGE)) {
					openDocument((String)message.get("title"),
								 (String)message.get("mime"),
								 (String)message.get("url"),
								 (Boolean)message.get("openable"),
								 (Boolean)message.get("grantWrite"),
								 (Boolean)message.get("persistPermissions"));
				} else if (type.equals(CREATE_DOCUMENT_MESSAGE)) {
					createDocument((String)message.get("title"),
								   (String)message.get("mime"),
								   (String)message.get("url"),
								   (String)message.get("name"),
								   (Boolean)message.get("openable"),
								   (Boolean)message.get("persistPermissions"));
				} else if (type.equals(SystemDispatcher.ACTIVITY_RESULT_MESSAGE)) {
					onActivityResult((Integer)message.get("requestCode"),
									 (Integer)message.get("resultCode"),
									 (Intent)message.get("data"));
				}
			}
		});
	}

	static private void getContent(final String title, String mime, boolean openable, boolean localOnly, boolean grantWrite) {
		final Activity activity = QtNative.activity();

		final Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
		intent.setType(mime);
		intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
		if(grantWrite)
			intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
		if(openable)
			intent.addCategory(Intent.CATEGORY_OPENABLE);
		if(localOnly)
			intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);

		Runnable runnable = new Runnable () {
			public void run() {
				activity.startActivityForResult(
					Intent.createChooser(intent, title),
					CHOOSE_CONTENT_ACTION);
			};
		};
		activity.runOnUiThread(runnable);
	}

	static private void openDocument(final String title, String mime, String url, boolean openable, boolean grantWrite, final boolean persistPermissions) {
		final Activity activity = QtNative.activity();

		final Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
		intent.setType(mime);
		intent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
		intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
		if(grantWrite)
			intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
		if(openable)
			intent.addCategory(Intent.CATEGORY_OPENABLE);
		//TODO add in Android O
//		if(!url.isEmpty())
//			intent.putExtra(Intent.EXTRA_INITIAL_URI, Uri.parse(url));

		Runnable runnable = new Runnable () {
			public void run() {
				activity.startActivityForResult(
					Intent.createChooser(intent, title),
					persistPermissions ? CHOOSE_PERSISTENT_ACTION : CHOOSE_CONTENT_ACTION);
			};
		};
		activity.runOnUiThread(runnable);
	}

	static private void createDocument(final String title, String mime, String url, String name, boolean openable, final boolean persistPermissions) {
		final Activity activity = QtNative.activity();

		final Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
		intent.setType(mime);
		intent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
		intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
		intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
		if(openable)
			intent.addCategory(Intent.CATEGORY_OPENABLE);
		//TODO add in Android O
//		if(!url.isEmpty())
//			intent.putExtra(Intent.EXTRA_INITIAL_URI, Uri.parse(url));
		if(!name.isEmpty())
			intent.putExtra(Intent.EXTRA_TITLE, name);

		Runnable runnable = new Runnable () {
			public void run() {
				activity.startActivityForResult(
					Intent.createChooser(intent, title),
					persistPermissions ? CHOOSE_PERSISTENT_ACTION : CHOOSE_CONTENT_ACTION);
			};
		};
		activity.runOnUiThread(runnable);
	}

	static private void onActivityResult(int requestCode, int resultCode, Intent data) {
		if(requestCode == CHOOSE_CONTENT_ACTION ||
		   requestCode == CHOOSE_PERSISTENT_ACTION) {
			Map reply = new HashMap();
			if (resultCode == Activity.RESULT_OK) {
				Uri uri = data.getData();

				if(requestCode == CHOOSE_PERSISTENT_ACTION) {
					int takeFlags = data.getFlags() & (Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
					QtNative.activity()
						.getContentResolver()
						.takePersistableUriPermission(uri, takeFlags);
				}

				reply.put("uri", uri.toString());
				reply.put("success", true);
			} else
				reply.put("success", false);
			SystemDispatcher.dispatch(CONTENT_CHOOSEN_MESSAGE, reply);
		}
	}
}
